Refactor/streamline block evaluation

greymistcube opened the pull request:
This PR attempts to resolve the problem mentioned in #1301.

Currently, what should be a relatively simple `BlockEvaluator.Evaluate()` call is handled in four different places. To mitigate this issue, a new `ActionEvaluator` class is introduced to keep the entire evaluation process within this new class. All methods for `ActionEvaluator` are pretty much copied from `BlockEvaluator`, `Block<T>`, `Transaction<T>`, and `ActionEvaluation`, aside from minor internal refactoring and renaming for better consistency and clarity.

limebell commented on CHANGES.md:
```diff
@@ -260,6 +270,7 @@ To be released.
 [#1289]: https://github.com/planetarium/libplanet/pull/1289
 [#1295]: https://github.com/planetarium/libplanet/issues/1295
 [#1296]: https://github.com/planetarium/libplanet/pull/1296
+[#1301]: https://github.com/planetarium/libplanet/issues/1301
```
```suggestion
[#1301]: https://github.com/planetarium/libplanet/issues/1301
[#1305]: https://github.com/planetarium/libplanet/pull/1305
```

greymistcube replied:
Yup, this isn't fully ready for deployment. :(

limebell reviewed (changes requested):
Some tests seem to fail because of the changes of this PR. Please check [Azure DevOps](https://dev.azure.com/planetarium/libplanet/_build/results?buildId=5108&view=results) for more detail.

greymistcube commented:
https://github.com/planetarium/libplanet/blob/25f62f8561c1f9037c1033ddf951d7a612030240/Libplanet.Tests/Blockchain/BlockChainTest.cs#L1826-L1837

This is one of the failing test, but I'm not sure what the intention is here. If we are evaluating a block action for the `genesis` block, shouldn't assertion be equal to `1`?

limebell commented:
Not a strong suggestion though, how about squash some commits? (i.e. [`414e8fd`](https://github.com/planetarium/libplanet/pull/1305/commits/414e8fd9d6eabb499a5c1df1dc273dd1765b66ee) and [`85710af`](https://github.com/planetarium/libplanet/pull/1305/commits/85710af8285457231d14155c765d3083256368b4))

greymistcube commented:
Should be ready for review. Since the changes are quite substantial, I'll be writing a summary soon.

greymistcube commented:
# Summary

New `ActionEvaluator<T>` class is introduced to handle all evaluation.

## Migration

- `ActionEvaluator<T>.Evaluate()`: Changed from `BlockEvaluator.EvaluateActions()`.
- `ActionEvaluator<T>.EvaluateBlock()`: Changed from `Block<T>.Evaluate()`.
- `ActionEvaluator<T>.EvaluateTxs()`: Changed from `Block<T>.EvaluateActionsPerTx()`.
- `ActionEvaluator<T>.EvaluateTx()`: Changed from `Transaction<T>.EvaluateActionsGradually()`.
- `ActionEvaluator<T>.EvaluateTxResult()`: Changed from `Transaction<T>.EvaluateActions()`.
- `ActionEvaluator<T>.EvaluatePolicyBlockAction()`: Changed from `BlockEvaluator.EvaluateBlockAction()`.
- `ActionEvaluator<T>.EvaluateActions()`: Changed from `ActionEvaluation.EvaluateActionsGradually()`.
- Other corresponding test methods moved to `ActionEvaluatorTest.cs`.

## Introduced

- `ActionEvaluator<T>.GetUpdatedAddresses()`: Separated from `ActionEvaluator<T>.EvaluateTxResult()`.
  - Used by `Transaction<T>` to get updated addresses.
  - Hard coded dependency injection from `Transaction<T>` removed. This is now handled internally.

## Notable changes

- Only the main entry point for evaluation, namely `ActionEvaluator<T>.Evaluate()`, directly takes external delegates.
  - All other methods depend on the previous states of type `IAccountStateDelta` before method call as evaluation context.
- Access modifier of `ActionEvaluator<T>.Evaluate()` has changed to `public` from `internal` and that of `ActionEvaluator<T>.EvaluateBlock()` has changed to `internal` from `public`.
- Number of parameters have been greatly reduced.
  - This is mainly done by requiring `Block<T>` as an argument instead of using individual values of `Block<T>`'s fields.
- Lazy evaluation is now strictly enforced for all methods with return type `IEnumerable<T>`.
  - Entire evaluation is postponed until right before returning the whole list of `ActionEvaluation`s from `ActionEvaluator<T>.Evaluate()` which has a return type of `IReadOnlyList<T>`.
  - This was not the case where a realization of `IEnumerable<T>` was prematurely happening inside `ActionEvaluator<T>.EvaluateBlock()`.
- `ActionEvaluator<T>.EvaluateTxs()` now has a simple `IEnumerable<Transaction<T>>` return type.
  - This is done by moving "updated addresses check" down to `ActionEvaluator<T>.EvaluateTx()` from `ActionEvaluator<T>.EvaluateBlock()`.
  - This removes the necessity for keeping track of which `Transaction<T>` an `ActionEvaluation` is tied to.
- A couple of private helper methods, `ActionEvaluator<T>.GetPreviousBlockOutputStates()` and `ActionEvaluator<T>.InitializeAccountGettersPair()`, are introduced to initialize the setup.

## Future suggestions

- A call to `Block<T>.Validate()` should not be inside `ActionEvaluator<T>.Evaluate()`.
- I suggest making `_policyBlockAction` non-nullable.
  - Designwise , it is probably better to have a default `IAction` that does nothing when evaluated.
- I also suggest making `_trieGetter` non-nullable.
  - Not even sure if trie data type should be enforced. Either way, if we want a mock storage, this should be handled outside the `ActionEvaluator<T>` class.
- Probably better to have parameters of `ActionEvaluator<T>.EvaluateActions` simplified. All the necessary information can be inferred from `Block<T>` and `Transaction<T>` to evaluate.
- The usage of `StateCompleterSet<T>` together with `StateGetter<T>`, `BalanceGetter<T>`, `AccountStateGetter`, and `AccountBalanceGetter` is quite convoluted. Concurrent refactoring may be needed.
  - Although I have introduced `StateGetter<T>` and `BalanceGetter<T>` as `public` `delegate`s for convenience, these `delegate`s are lacking documentation. I do not think I fully understand the purpose of these, so I can't concisely explain what these are for.
- One additional miner qualm is that `Block<T>.Miner` is nullable at the moment. It is probably better to be able to use `Block<T>.Miner` instead of `Block<T>.Miner.GetValueOrDefault()` from the perspective of the `ActionEvaluator<T>` class.

## Remarks

- Substantial amount of documentation has been reworked.
- I did my best to fix the tests to accommodate the changes made, but due to my lack of experience in unit testing, updated test code may need extra care while reviewing.
- I am sure there are other things I forgot to mention. I'll make additional comments as things come up.

longfin commented:
/rebase

limebell commented:
![image](https://user-images.githubusercontent.com/5073795/120412301-4cbc1180-c391-11eb-9f48-8fd00e85295c.png)

Build seems to fail in mono.

greymistcube commented:
@limebell 
Fixed. Older language versions not supporting `Zip()` taking a single argument was the issue.

longfin commented on Libplanet/Action/ActionEvaluator.cs:
```diff
@@ -0,0 +1,581 @@
+                rehearsal: true,
+                previousBlockStatesTrie: null);
+
+            if (evaluations.Any())
+            {
+                return evaluations.Last().OutputStates.UpdatedAddresses;
+            }
```
Trivial
```suggestion
            if (evaluations.LastOrDefault() is { } lastEval)
            {
                return lastEval.OutputStates.UpdatedAddresses;
            }
```

greymistcube replied:
Hmm... haven't really put my thought into it when writing this, but looking back, wouldn't this be problematic if for whatever reason `default` is implemented and `AccountStateDeltaImpl` becomes a `struct`? I know it is clear from the code that `AccountStateDeltaImpl` is a `class` (and also it implements an `interface`, which also makes it "likely" to be a `class`) but conceptually, I don't think an evaluation output should necessarily be a `class`.

longfin replied:
Fair enough.

longfin commented on Libplanet/Action/ActionEvaluator.cs:
```diff
@@ -0,0 +1,581 @@
+            IAccountStateDelta previousStates,
+            bool rehearsal = false)
+        {
+            var evaluations = EvaluateTx(
```
```suggestion
            ImutableList<ActionEvaluation> evaluations = EvaluateTx(
```

longfin commented on Libplanet/Action/ActionEvaluator.cs:
```diff
@@ -0,0 +1,581 @@
+                actions: new[] { _policyBlockAction }.ToImmutableList(),
+                rehearsal: false,
+                previousBlockStatesTrie: previousBlockStatesTrie,
+                blockAction: true).First();
```
```suggestion
                blockAction: true).Single();
```

greymistcube commented on Libplanet/Action/ActionEvaluator.cs:
```diff
@@ -0,0 +1,581 @@
+            IAccountStateDelta previousStates,
+            bool rehearsal = false)
+        {
+            ImutableList<ActionEvaluation> evaluations = EvaluateTx(
```
```suggestion
            ImmutableList<ActionEvaluation> evaluations = EvaluateTx(
```
Typo.

dahlia commented on Libplanet/Action/ActionEvaluator.cs:
```diff
@@ -0,0 +1,581 @@
+                ^ (signature.Any() ? BitConverter.ToInt32(hashedSignature, 0) : 0);
+
+            IAccountStateDelta states = previousStates;
+            ILogger logger = Log.ForContext<ActionEvaluation>();
```
Why not use `_logger` instead?

greymistcube replied:
Partly because `EvaluateActions()` is left as `static`. Not sure if it was the best design choice, but the my reasoning behind it was to leave last mile/heavy lifting as `static` since this may be considered more of as a utility method, and possibly for optimization.

dahlia replied:
As logger contexts are for filtering logs, to me the best practice is to make logs belong to the context where the logging code is placed file-wise.  The current context implies the logging code is placed in *ActionEvaluation.cs*, it would be rather a trap when we investigate issues in production.

greymistcube replied:
Well, I'll just turn the method into an instance method. Shouldn't affect the performance too hard anyways.

I think a more coherent solution would be to make it `private` and handle it from outside.

greymistcube replied:
Eh... ended up making `_logger` static instead. See 87849869dbdab8c8548cc56db3e054bd1dd291d9.

dahlia commented on Libplanet.Tests/Action/ActionEvaluatorTest.cs:
```diff
@@ -0,0 +1,997 @@
+            var stateRootBlock = TestUtils.MineGenesis(
+                timestamp: timestamp,
+                transactions: txs).AttachStateRootHash(stateStore, null);
+            StateGetter<RandomAction> nullStateGetter =
+                (address, hashDigest, stateCompleter) => null;
+            BalanceGetter<RandomAction> nullBalanceGetter =
+                (address, currency, hashDigest, fungibleAssetStateCompleter)
+                    => new FungibleAssetValue(currency);
```
Would such null state/balance getters be frequently used?  If so, how about changing the visibility of `ActionEvaluator<T>._nullStateGetter` & `_nullBalanceGetter` into `public static readonly`?  (Or it might be better to define them in normal static methods instead.  If a method's signature is matched to a delegate type-wise, the compile treat the method as a value of the delegate type.)

greymistcube replied:
Went back and forth with this a little bit. Thought about exposing `ActionEvaluator<T>._nullStateGetter` and such as `public`, but then realized it would most likely only be for testing (and possibly for benchmarking). Now that I'm going over `ActionEvaluator.cs` again, `ActionEvaluator<T>._nullStateGetter` and `ActionEvaluator<T>._nullBalanceGetter` don't even get used. 🙃

I'll just remove them from `ActionEvaluator<T>`.

dahlia replied:
If these are not useful to Libplanet users, but still useful for test/benchmark code how about turning them `internal static readonly`?

greymistcube replied:
Hmm... I'm not fully on board with the idea of making things `internal` for purely testing and/or benchmarking. However, from a slightly different point of view, I can get behind the idea of providing additional `internal` tools for developers using the library, including us, of course. I'll make the changes. 🙃

greymistcube replied:
On a second thought, I still think the answer is no. I'll make the changes regardless for now, since this can be separated to a different issue.

greymistcube replied:
Done. See 04ae033430c70f543115322e3f613da1a3d8a113.

dahlia commented on Libplanet.Tests/Action/ActionEvaluatorTest.cs:
```diff
@@ -0,0 +1,997 @@
+            var actionEvaluation = chain.ActionEvaluator.Evaluate(
+                chain.Tip,
+                StateCompleterSet<TestEvaluateAction>.Recalculate);
+            Assert.False(actionEvaluation[0].InputContext.BlockAction);
```
Could you test if `ActionEvaluation`'s other properties have correct values too?  Especially, it would be better if it properly has `Exception` on failed actions.

dahlia replied:
@greymistcube It seems this is not addressed yet.

greymistcube replied:
Still working on it. 😢

greymistcube replied:
Implemented with mostly copy pasted code. See 55e4ec57c71953cead85c364719d3832aa5fad56.

dahlia reviewed (commented):
> - I suggest making `_policyBlockAction` non-nullable.
>   - Designwise , it is probably better to have a default `IAction` that does nothing when evaluated.

On .NET, interfaces cannot have `default` values, because they are treated as reference types.  Their default values are merely `null`.

> - The usage of `StateCompleterSet<T>` together with `StateGetter<T>`, `BalanceGetter<T>`, `AccountStateGetter`, and `AccountBalanceGetter` is quite convoluted. Concurrent refactoring may be needed.

If it's possible I believe simply adding parameters to the existing `AccountStateGetter` & `AccountBalanceGetter` delegates would be easier to understand and maintain than having one more pair of `StateGetter<T>` & `BalanceGetter<T>`.  Of course, they need to have generic parameters (`AccountStateGetter<T>` & `AccountBalanceGetter<T>`) too if we'd like to do so.

---

Good job.  Thanks for your hard work.  Anyway, your branch is now outdated.  Needs to rebase again.

greymistcube commented on Libplanet/Action/AccountBalanceGetter.cs:
```diff
@@ -16,4 +18,11 @@ namespace Libplanet.Action
     /// </returns>
     public delegate FungibleAssetValue AccountBalanceGetter(Address address, Currency currency);
+
+    public delegate FungibleAssetValue BalanceGetter<T>(
+        Address address,
+        Currency currency,
+        BlockHash? hashDigest,
+        FungibleAssetStateCompleter<T> fungibleAssetStateCompleter)
+            where T : IAction, new();
```
Sorry, I got bit lazy towards the end. Didn't really decide what to do with `delegates`. After a brief survey, the consensus seems to be to either put them in the related class file, which would be `ActionEvaluator.cs` in this case, or have separate files, which seems bit excessive. In any case, I'll separate them out.

greymistcube replied:
See cc305a5fbf1c1da20980b58283c715f9efb92e6b.

greymistcube commented on Libplanet/Action/ActionEvaluator.cs:
```diff
@@ -0,0 +1,581 @@
+            AccountStateGetter accountStateGetter;
+            AccountBalanceGetter accountBalanceGetter;
+
+            if (block.PreviousHash is null)
+            {
+                accountStateGetter = _nullAccountStateGetter;
+                accountBalanceGetter = _nullAccountBalanceGetter;
+            }
+            else
+            {
+                accountStateGetter = address => _stateGetter(
+                    address,
+                    block.PreviousHash,
+                    stateCompleterSet.StateCompleter);
+                accountBalanceGetter = (address, currency) => _balanceGetter(
+                    address,
+                    currency,
+                    block.PreviousHash,
+                    stateCompleterSet.FungibleAssetStateCompleter);
+            }
```
~Empty braces non-null pattern matching weirds me out.~

I'll fix it.

greymistcube commented:
> On .NET, interfaces cannot have default values, because they are treated as reference types. Their default values are merely null.

I disagree. I was thinking something akin to `_nullStateGetter` and such. I know the language does not support `default` values for `interface`es or `class`es per se, but I don't think this means we can't have a default behavior. To drive my point, I would say something like

```csharp
int Sum(List<int>? list)
{
    if (list is { })
    {
        return list.Sum();
    }
    else
    {
        return 0;
    }
}
```

is, in general, a bad design choice. A `null` check should not be deferred to an operating entity, but handled by the caller or, better yet, the list creator. Besides, `IBlockPolicy<T>.BlockAction` is not supposed to be `null`, but it just so happens `BlockPolicy<T>.BlockAction` implementation can be, due to discrepancy of nullable contexts, which should be fixed anyway.

Also, there is already a `NullAction` class found in `StoreExtensions.cs`. 🙃

greymistcube commented on Libplanet/Action/ActionEvaluator.cs:
```diff
@@ -0,0 +1,581 @@
+    public class ActionEvaluator<T>
+        where T : IAction, new()
+    {
+        private static readonly StateGetter<T> _nullStateGetter =
+            (address, hashDigest, stateCompleter) => null;
+
+        private static readonly BalanceGetter<T> _nullBalanceGetter =
+            (address, currency, hashDigest, fungibleAssetStateCompleter)
+                => new FungibleAssetValue(currency);
+
+        private static readonly AccountStateGetter _nullAccountStateGetter = address => null;
+        private static readonly AccountBalanceGetter _nullAccountBalanceGetter =
+            (address, currency) => new FungibleAssetValue(currency);
```
```suggestion
        private static readonly AccountStateGetter _nullAccountStateGetter = address => null;
        private static readonly AccountBalanceGetter _nullAccountBalanceGetter =
            (address, currency) => new FungibleAssetValue(currency);
```

Remove unused code.

dahlia commented:
> I disagree. I was thinking something akin to `_nullStateGetter` and such. I know the language does not support `default` values for `interface`s or `class`es per se, but I don't think this means we can't have a default behavior.

Fair enough, but we still don't need to fill it with a null getter by default.  If we make the constructor parameter required callers should explicitly pass a null getter by their hands.  To me, this way is a more error-prone design.

greymistcube commented on Libplanet/Action/ActionEvaluator.cs:
```diff
@@ -0,0 +1,581 @@
+
+namespace Libplanet.Action
+{
+    public class ActionEvaluator<T>
```
See 0b854364be609ed868e6422f385739bee6a6120a. Might be tad bit short. 😶

longfin reviewed (dismissed):
But another conflict appears 😢
